scripts/esbuild/util.ts

```diff
@@ -0,0 +1,88 @@
+import { join } from 'path';
+import { BuildOptions } from '../utils/options';
+
+/**
+ * Aliases which map the module identifiers we set in `paths` in `tsconfig.json` to
+ * bundles (built either with esbuild or with rollup).
+ *
+ * @param opts options for the current build
+ * @returns a map of aliases to bundle entry points, relative to the root directory
+ */
+export function getEsbuildAliases(opts: BuildOptions): Record<string, string> {
+  return {
+    // node module redirection
+    chalk: 'ansi-colors',
+
+    // mapping aliases to top-level bundle entrypoints
+    '@stencil/core/testing': './testing/index.js',
+    '@stencil/core/compiler': './compiler/stencil.js',
+    '@stencil/core/dev-server': './dev-server/index.js',
+    '@stencil/core/mock-doc': './mock-doc/index.cjs',
+    '@stencil/core/internal/testing': './internal/testing/index.js',
+
+    // this has to be added here pointing to the `.ts` file so that the `import
+    // { IS_NODE_ENV } from '@environment';` in the `typescript.js` file we
+    // bundle will resolve correctly
+    //
+    // esbuild will automatically resolve aliases from `paths` in
+    // `tsconfig.json` but only for `.ts` files, so when we bundle `typescript`
+    // into e.g. the compiler bundle we'll need this so things resolve
+    // correctly
+    '@environment': join(opts.srcDir, 'compiler/sys/environment.ts'),
+  };
+}
+
+/**
+ * Node modules which should be universally marked as external
+ *
+ */
+const externalNodeModules = [
```

Comment:
the CLI module doesn't pull most of these in, but they are used by other bundles

Reply:
it may not be totally exhaustive - on my other branch this was the external modules I collated while getting the `cli`, `compiler`, `testing`, and `mock-doc` bundles building with Esbuild, so when we go to tackle the others maybe we'll need a few more!

Reply:
Man, I really wish there was a way to automatically generate this list (and the list of module aliases above). Feels like something easy to overlook.

Did you come up with all these by trial and error (like, does the build fail if we're missing one of these), or did you have to scan the code for each bundle to make sure you had all of them?

Reply:
I based this on the existing `external` lists used for the rollup builds (see [here](https://github.com/ionic-team/stencil/blob/a0dc40512c47982c1f933f59f71206cef4d1923c/scripts/bundles/testing.ts#L33-L64), [here](https://github.com/ionic-team/stencil/blob/a0dc40512c47982c1f933f59f71206cef4d1923c/scripts/bundles/cli.ts#L66-L66), and [here](https://github.com/ionic-team/stencil/blob/a0dc40512c47982c1f933f59f71206cef4d1923c/scripts/bundles/screenshot.ts#L38-L38) for instance) and then I think possibly had to add one or two things that weren't already present - it is a bit of an annoying thing, if we accidentally forget to add something we do want to have externalized to the list then esbuild (and rollup too!) will happily just pull it into the bundle.